uroute/url.py

```diff
@@ -67,31 +67,37 @@ def clean_url(self, url):
             if not re.match(provider['urlPattern'], url, re.IGNORECASE):
                 continue
 
+            # If any exceptions are matched, this provider is skipped
             if any(
                 re.match(exc, url, re.IGNORECASE)
-                for exc in provider['exceptions']
+                for exc in provider.get('exceptions', [])
```

Comment:
Switching from the usual key lookup syntax to the `get()` method allows for default values, which enables processing of sparsely defined rules (i.e. rules with incomplete schemas). I think this is useful because it allows custom rules to be written lazily.